compiler/typecheck/TcHsType.hs

```diff
@@ -1472,7 +1472,7 @@ kcHsTyVarBndrs name flav cusk all_kind_vars
       = tcExtendTyVarEnv [tv] thing_inside
 
     kc_hs_tv :: HsTyVarBndr GhcRn -> TcM (TcTyVar, Bool)
-    kc_hs_tv (UserTyVar lname@(L _ name))
+    kc_hs_tv (UserTyVar lname@(L _ name) _)
```

Comment:
Ah, okay, that clears up a misconception I had: I'd thought that any `HsTyVarBndr` had to contain an `ArgFlag Invisible`—but there are circumstances in which it might be `Visible`? This only shows up in data constructors, is that correct?

If so, I like this idea of percolating the index up into the `HsTyVarBndr` as well, and naturally to `LHsTyVarBndr` (and probably a few other places).

It also occurs to me: these indices, to match, should show up in TH too. It's okay to use GADTs in the TH API, right?

Reply:
I had that same misconception. However, it's not true for type declarations: things like `data T a b c` or `type family F a b c`. So it's not *data* constructors, but *type* constructors.

I would avoid GADTs in TH. TH will just be over-expressive. The TH datatypes have `Data` and `Generic` instances, which people make good use of.